mobile/src/main/java/rs/readahead/washington/mobile/views/activity/VideoViewerActivity.kt

```diff
@@ -0,0 +1,540 @@
+        // If the event was not handled then see if the player view can handle it as a media key event.
+        return super.dispatchKeyEvent(event) || simpleExoPlayerView.dispatchMediaKeyEvent(event)
+    }
+
+    private fun shareMediaFile() {
+        if (vaultFile == null) {
+            return
+        }
+        if (vaultFile?.metadata != null) {
+            showShareWithMetadataDialog()
+        } else {
+            startShareActivity(false)
+        }
+    }
+
+    private fun startShareActivity(includeMetadata: Boolean) {
+        if (vaultFile == null) {
+            return
+        }
+        MediaFileHandler.startShareActivity(this, vaultFile, includeMetadata)
+    }
+
+    private fun initializePlayer() {
+        val needNewPlayer = player == null
+        if (needNewPlayer) {
+            player = ExoPlayer.Builder(this).build()
+            player?.apply {
+                binding.playerView.player
+                playWhenReady = shouldAutoPlay
+            }
+
+            simpleExoPlayerView.setPlayer(player)
+        }
+        if (needNewPlayer || needRetrySource) {
+            if (intent.hasExtra(VIEW_VIDEO) && intent.extras != null) {
+                val vaultFile = intent.extras!![VIEW_VIDEO] as VaultFile?
+                if (vaultFile != null) {
+                    this.vaultFile = vaultFile
+                    toolbar?.title = vaultFile.name
+                    setupMetadataMenuItem(vaultFile.metadata != null)
+                }
+            }
+            val mediaFileDataSourceFactory = MediaFileDataSourceFactory(
+                this, vaultFile!!, null
+            )
+
+            val mediaItem = MediaItem.fromUri(MediaFileHandler.getEncryptedUri(this, vaultFile))
+            val mediaSource: MediaSource =
+                ProgressiveMediaSource.Factory(mediaFileDataSourceFactory)
+                    .createMediaSource(mediaItem)
+
+            val haveResumePosition = resumeWindow != C.INDEX_UNSET
+            if (haveResumePosition) {
+                player!!.seekTo(resumeWindow, resumePosition)
+            }
+            player!!.prepare(mediaSource, !haveResumePosition, false)
+            needRetrySource = false
+        }
+    }
+
+    private fun releasePlayer() {
+        if (player != null) {
+            shouldAutoPlay = player?.playWhenReady == true
+            //updateResumePosition(); // todo: fix source skipping..
+            player?.release()
+            player = null
+            trackSelector = null
+            clearResumePosition()
+        }
+    }
+
+    private fun clearResumePosition() {
+        resumeWindow = C.INDEX_UNSET
+        resumePosition = C.TIME_UNSET
+    }
+
+    override fun onVisibilityChange(visibility: Int) {
+        if (!isInfoShown) {
+            toolbar!!.visibility = visibility
+        } else {
+            toolbar!!.visibility = View.VISIBLE
+        }
+    }
+
+    private fun setupToolbar() {
+        toolbar = binding.playerToolbar
```

Comment:
toolbar should be `lateinit` also or we can access it directly from the binding so we don't have to use **not-null assertion operator**.